### PR TITLE
Added MapController.pointToLatLng() to get LatLng of given screen point

### PR DESCRIPTION
**Pull request by HugoHeneault**
I wanted to get LatLng of a given point on screen, but couldn't find an easy way to. 

So I've implemented a `MapController.pointToLatLng()` inspired by @maRci002's [answer](https://github.com/fleaflet/flutter_map/issues/607#issuecomment-628208311).

Ref #496, ref #607, ref #981, ref #1010

### CODE REVIEW

**Comment by HugoHeneault** [comment]
Here's an example of what it does:
simulator_screenshot_6240A1E0-F7F7-48B6-9434-254F3B65F568

**Comment by JaffaKetchup** [comment]
Fantastic! I recently manually coded this using parts from the library, but I couldn't do it easily because I didn't have access to internal variables.
This should make things much easier! Thanks.

**Review by ibrierley — COMMENTED** [review]
I like this, does it pass dartfmt ok  ? (not sure what we're using these days, but line 86 indentation map.dart looks like it should be spaced out, but this is very minor)?

**Comment by HugoHeneault** [comment]
> I like this, does it pass dartfmt ok ? (not sure what we're using these days, but line 86 indentation map.dart looks like it should be spaced out, but this is very minor)?

Nope sorry, I'll `dart format` 

Is there a way to get it done automatically?

**Comment by HugoHeneault** [comment]
I just pushed a fix that caused calculation to be wrong when user zoomed the map. 
Now I have an issue when user rotates the map. 

https://user-images.githubusercontent.com/4112256/149125971-a724c28f-9038-4ea4-91de-cfe1ec820019.mov

Has anyone an idea?

**Comment by HugoHeneault** [comment]
I updated my example but still can't figure out what's wrong :(

It looks like `LatLng` I return doesn't take rotation in consideration. Still investigating, any help would be nice :)

https://user-images.githubusercontent.com/4112256/149280272-51bd9ebe-9c09-47a7-af09-89a74f702f65.mp4

**Comment by ibrierley** [comment]
Are you still having problems with the rotation issue ?

**Comment by HugoHeneault** [comment]
Yes no updates here. Might need some help... We've - for now - disabled rotation on our app when this is active.

**Comment by ibrierley** [comment]
See if this works for you....
````
@override
  LatLng? pointToLatLng(CustomPoint localPoint) {
    if (_state.originalSize == null) {
      return null;
    }

    final width = _state.originalSize!.x;
    final height = _state.originalSize!.y;

    var localPointCenterDistance =
        CustomPoint((width / 2) - localPoint.x, (height / 2) - localPoint.y);
    var mapCenter =
        _state.options.crs.latLngToPoint(_state.center, _state.zoom);

    var point = mapCenter - localPointCenterDistance;

    if( _state.rotation != 0.0 ) {
      var m = Matrix4.identity()
        ..translate(mapCenter.x.toDouble(), mapCenter.y.toDouble())
        ..rotateZ(-_state.rotationRad)
        ..translate(-mapCenter.x.toDouble(), -mapCenter.y.toDouble());

      var tp = MatrixUtils.transformPoint(
          m, Offset(point.x.toDouble(), point.y.toDouble()));

      return _state.options.crs.pointToLatLng(
          CustomPoint(tp.dx, tp.dy), _state.zoom);
    }

    return _state.options.crs.pointToLatLng(
        point, _state.zoom);
  }
```

If it does (the icon rotates around its own center, but I think either a nonrotatedwidget or an opposite rotation would probably fix), maybe the rotated matrix stuff could be broken out into a separate method, as there may be uses for it being called elsewhere....

**Comment by ibrierley** [comment]
Just wondering on this one, if pointToLatLng should take the option of using the maps rotation (then it could use the matrix code). As some may want to use this as a separate method sometimes for a straightforward calculation ignoring rotation.

This also needs pull upstream so we can merge whenever, as there have been some changes since.

**Comment by HugoHeneault** [comment]
I'll be on vacation for two weeks, might work back on it around mi-february. Feel free to work on it during this time :)
Thanks @ibrierley

**Comment by JaffaKetchup** [comment]
@HugoHeneault Hope you enjoyed your vacation. Any news on this? Thanks.

**Comment by HugoHeneault** [comment]
No and I don't have time this days so if you'd like to investigate do not hesitate :)

**Comment by HugoHeneault** [comment]
This should still be reviewed and fixed as I'm sure other people will need it...

**Comment by pablojimpas** [comment]
@HugoHeneault would you mind re-basing this branch from master when you have time? I'll take a look at it after that.

**Comment by HugoHeneault** [comment]
> @HugoHeneault would you mind re-basing this branch from master when you have time? I'll take a look at it after that.

I'm on it :)

**Comment by HugoHeneault** [comment]
Rebase done ✅

File: lib/src/map/map.dart [hunk 1]
```diff
@@ -78,6 +78,24 @@ class MapControllerImpl implements MapController {
...
+
+    final localPointCenterDistance =
+        CustomPoint((width / 2) - localPoint.x, (height / 2) - localPoint.y);
+    final mapCenter =
+        _state.options.crs.latLngToPoint(_state.center, _state.zoom);
+
+    final point = mapCenter - localPointCenterDistance;
+    return _state.options.crs.pointToLatLng(point, _state.zoom);
+  }
+
```
Comment (pablojimpas) [inline lib/src/map/map.dart:98]: I've tested this with the suggestion proposed by @ibrierley and the rotation issue seems to be fixed AFAIK.

```suggestion
@override
  LatLng? pointToLatLng(CustomPoint localPoint) {
    if (_state.originalSize == null) {
      return null;
    }

    final width = _state.originalSize!.x;
    final height = _state.originalSize!.y;

    final localPointCenterDistance =
        CustomPoint((width / 2) - localPoint.x, (height / 2) - localPoint.y);
    final mapCenter =
        _state.options.crs.latLngToPoint(_state.center, _state.zoom);

    var point = mapCenter - localPointCenterDistance;

    if (_state.rotation != 0.0) {
      point = rotatePoint(mapCenter, point);
    }

    return _state.options.crs.pointToLatLng(point, _state.zoom);
  }

  CustomPoint<num> rotatePoint(
      CustomPoint<num> mapCenter, CustomPoint<num> point) {
    final m = Matrix4.identity()
      ..translate(mapCenter.x.toDouble(), mapCenter.y.toDouble())
      ..rotateZ(-_state.rotationRad)
      ..translate(-mapCenter.x.toDouble(), -mapCenter.y.toDouble());

    final tp = MatrixUtils.transformPoint(
        m, Offset(point.x.toDouble(), point.y.toDouble()));

    return CustomPoint(tp.dx, tp.dy);
  }
```

**Comment by JaffaKetchup** [comment]
Ok, thanks @pablojimpas. Will commit those suggestions to the PR (I think @HugoHeneault has allowed maintainer edits, thanks). Then we can try testing again :)

**Comment by JaffaKetchup** [comment]
Annoying. Will have to find a way to contribute directly if possible.

**Comment by pablojimpas** [comment]
> Annoying. Will have to find a way to contribute directly if possible.

I'm not sure if you (as a maintainer) can push commits to this branch directly from your local machine, maybe only the owner is allowed.

**Comment by JaffaKetchup** [comment]
I do have write access which I think is what counts, but you might be right as I've never done it before. Can't do it right now, should be able to do it in about an hour's time.

**Comment by JaffaKetchup** [comment]
Yeah, I can't do anything directly as it's someone else's fork. I'll try fixing it through suggestions...

File: lib/src/map/map.dart [hunk 2]
```diff
@@ -78,6 +78,41 @@ class MapControllerImpl implements MapController {
...
+    final m = Matrix4.identity()
+      ..translate(mapCenter.x.toDouble(), mapCenter.y.toDouble())
+      ..rotateZ(-_state.rotationRad)
+      ..translate(-mapCenter.x.toDouble(), -mapCenter.y.toDouble());
+
+    final tp = MatrixUtils.transformPoint(
+        m, Offset(point.x.toDouble(), point.y.toDouble()));
+
+    return CustomPoint(tp.dx, tp.dy);
+  }
```
Comment (JaffaKetchup) [inline lib/src/map/map.dart:115]: Trying to appease the checker...

```suggestion
LatLng? pointToLatLng(CustomPoint localPoint) {
  if (_state.originalSize == null) {
    return null;
  }

  final width = _state.originalSize!.x;
  final height = _state.originalSize!.y;

  final localPointCenterDistance =
      CustomPoint((width / 2) - localPoint.x, (height / 2) - localPoint.y);
  final mapCenter =
      _state.options.crs.latLngToPoint(_state.center, _state.zoom);

  var point = mapCenter - localPointCenterDistance;

  if (_state.rotation != 0.0) {
    point = rotatePoint(mapCenter, point);
  }

  return _state.options.crs.pointToLatLng(point, _state.zoom);
}

CustomPoint<num> rotatePoint(
    CustomPoint<num> mapCenter, CustomPoint<num> point) {
  final m = Matrix4.identity()
    ..translate(mapCenter.x.toDouble(), mapCenter.y.toDouble())
    ..rotateZ(-_state.rotationRad)
    ..translate(-mapCenter.x.toDouble(), -mapCenter.y.toDouble());

  final tp = MatrixUtils.transformPoint(
      m, Offset(point.x.toDouble(), point.y.toDouble()));

  return CustomPoint(tp.dx, tp.dy);
}
```

File: lib/src/map/map.dart [hunk 3]
```diff
@@ -78,6 +78,41 @@ class MapControllerImpl implements MapController {
...
+  final m = Matrix4.identity()
+    ..translate(mapCenter.x.toDouble(), mapCenter.y.toDouble())
+    ..rotateZ(-_state.rotationRad)
+    ..translate(-mapCenter.x.toDouble(), -mapCenter.y.toDouble());
+
+  final tp = MatrixUtils.transformPoint(
+      m, Offset(point.x.toDouble(), point.y.toDouble()));
+
+  return CustomPoint(tp.dx, tp.dy);
+}
```
Comment (JaffaKetchup) [inline lib/src/map/map.dart:115]: Maybe now?

```suggestion
  @override
  LatLng? pointToLatLng(CustomPoint localPoint) {
    if (_state.originalSize == null) {
      return null;
    }
  
    final width = _state.originalSize!.x;
    final height = _state.originalSize!.y;
  
    final localPointCenterDistance =
        CustomPoint((width / 2) - localPoint.x, (height / 2) - localPoint.y);
    final mapCenter =
        _state.options.crs.latLngToPoint(_state.center, _state.zoom);
  
    var point = mapCenter - localPointCenterDistance;
  
    if (_state.rotation != 0.0) {
      point = rotatePoint(mapCenter, point);
    }
  
    return _state.options.crs.pointToLatLng(point, _state.zoom);
  }
  
  CustomPoint<num> rotatePoint(
      CustomPoint<num> mapCenter, CustomPoint<num> point) {
    final m = Matrix4.identity()
      ..translate(mapCenter.x.toDouble(), mapCenter.y.toDouble())
      ..rotateZ(-_state.rotationRad)
      ..translate(-mapCenter.x.toDouble(), -mapCenter.y.toDouble());
  
    final tp = MatrixUtils.transformPoint(
        m, Offset(point.x.toDouble(), point.y.toDouble()));
  
    return CustomPoint(tp.dx, tp.dy);
  }
  
```

**Review by JaffaKetchup — COMMENTED** [review]
Maybe now?

**Comment by HugoHeneault** [comment]
If it's working, I'll apply the suggested fix tomorrow morning with formating, if you haven't figured it out already. Thanks guys for your help (I'm sorry I didn't followed the suggestion as I've moved on another project)

**Comment by JaffaKetchup** [comment]
@HugoHeneault Thanks. All is needed now (I think) is just the addition of a trailing newline. Should just be able to run `dart format .`, or do it manually. Don't worry about doing other stuff!

**Comment by HugoHeneault** [comment]
I did the reformat and tested. It works! Thanks alot @ibrierley for the solution and @pablojimpas & @JaffaKetchup for the follow-up!

LGTM :)

**Comment by JaffaKetchup** [comment]
Seems good to me, thanks to:

- @HugoHeneault 
- @pablojimpas 
- @ibrierley 

Will merge now...

**Review by JaffaKetchup — APPROVED** [review]
LGTM!

**Comment by Ronjack** [comment]
wow, just in time, perfect, thanks!

**Comment by mohammedX6** [comment]
> See if this works for you....
> 
> ```
> @override
>   LatLng? pointToLatLng(CustomPoint localPoint) {
>     if (_state.originalSize == null) {
>       return null;
>     }
> 
>     final width = _state.originalSize!.x;
>     final height = _state.originalSize!.y;
> 
>     var localPointCenterDistance =
>         CustomPoint((width / 2) - localPoint.x, (height / 2) - localPoint.y);
>     var mapCenter =
>         _state.options.crs.latLngToPoint(_state.center, _state.zoom);
> 
>     var point = mapCenter - localPointCenterDistance;
> 
>     if( _state.rotation != 0.0 ) {
>       var m = Matrix4.identity()
>         ..translate(mapCenter.x.toDouble(), mapCenter.y.toDouble())
>         ..rotateZ(-_state.rotationRad)
>         ..translate(-mapCenter.x.toDouble(), -mapCenter.y.toDouble());
> 
>       var tp = MatrixUtils.transformPoint(
>           m, Offset(point.x.toDouble(), point.y.toDouble()));
> 
>       return _state.options.crs.pointToLatLng(
>           CustomPoint(tp.dx, tp.dy), _state.zoom);
>     }
> 
>     return _state.options.crs.pointToLatLng(
>         point, _state.zoom);
>   }
> ```
> 
> If it does (the icon rotates around its own center, but I think either a nonrotatedwidget or an opposite rotation would probably fix), maybe the rotated matrix stuff could be broken out into a separate method, as there may be uses for it being called elsewhere....

Hi @ibrierley 
I want to take consideration the rotation if I want to convert from LatLng to point, How I can do it ? any hints please.

**Comment by ibrierley** [comment]
Are you all sorted on this now (I noted your other issue is closed), or still having an issue ?

**Comment by mohammedX6** [comment]
@ibrierley Thanks ,i got it working now.

**Comment by mohammedX6** [comment]
> Are you all sorted on this now (I noted your other issue is closed), or still having an issue ?

Hi @ibrierley I am sorry to distrubte you, I am trying to achieve this for days but without luck, i am trying here to convert LatLng Point (I need the point be a LatLng point so Local/Global position wont work) to a point (Offset)
Here what i am trying 
```
 @override
  Offset? LatLngToPoint(LatLng localPoint) {
    if (_state.originalSize == null) {
      return null;
    }

    final width = _state.originalSize!.x;
    final height = _state.originalSize!.y;

    var myLocalPoint =
        _state.options.crs.latLngToPoint(localPoint, _state.zoom);

    final localPointCenterDistance = CustomPoint(
        (width / 2) - myLocalPoint.x, (height / 2) - myLocalPoint.y);
    final mapCenter =
        _state.options.crs.latLngToPoint(_state.center, _state.zoom);

    var point = mapCenter - localPointCenterDistance;

    if (_state.rotation != 0.0) {
      point = rotatePoint(mapCenter, point);
      
    }

    return Offset(myLocalPoint.x.toDouble(), myLocalPoint.y.toDouble());
  } 
```
can you help me please

**Comment by ibrierley** [comment]
Why do you need the latlng from screen point out of interest, normally there is a simpler way.

**Comment by mohammedX6** [comment]
@ibrierley Thanks for response, I am using [flutter_map_line_editor](https://github.com/ibrierley/flutter_map_line_editor) `onDragUpdate` callback , using the `screenPoint.global` This is for touch position not for the actual point position, i want the actual point position because touch position and point position are not the same.
Screenshot (6)

**Comment by ibrierley** [comment]
Why are your touch and point positions not the same ?

**Comment by ibrierley** [comment]
From the info provided, I can't really see what you need to have working beyond the original code, so I'm probably missing something. 

If you want to provide a minimal example I can run, to highlight your issue, I will take a look and see if I can help.

**Comment by mohammedX6** [comment]
can we chat on the discord channel to make things easier , I will give in short time an example

**Comment by mohammedX6** [comment]
@ibrierley  Hi , Here is a minimal example https://filebin.net/f3na2ig0nilvp4hp/example.zip

**Comment by ibrierley** [comment]
You need to create a smaller single example, without any plugins and highlight what is wrong with pointToLatLng with a rotation.

**Comment by mohammedX6** [comment]
@ibrierley  There is no wrong with pointToLatLng, I just want LatLngToPoint with rotation

**Comment by ibrierley** [comment]
Ok, this issue is about pointToLatLng, and the code you quoted is that :D. Probably why the confusion with no example. 

Can you open a new issue that is latLngToPoint, not pointToLatLng, so it's clear what you need.

**Comment by mohammedX6** [comment]
@ibrierley  yes sure

**Pull request merged**